File: Hashing/java/pairProduct.java
import java.util.HashMap;
import java.util.Map;

class PairProduct {
  public static int[] pairProduct(int[] numbers, int targetProduct) {
    Map<Integer, Integer> indices = new HashMap<>();

    for (int i = 0; i < numbers.length; i += 1) {
      int current = numbers[i];
      if (current == 0) {
        if (targetProduct == 0 && indices.containsKey(0)) {
          return new int[] { indices.get(0), i };
        }
        indices.putIfAbsent(0, i);
        continue;
      }

      if (targetProduct % current != 0) {
        indices.putIfAbsent(current, i);
        continue;
      }

      int complement = targetProduct / current;
      if (indices.containsKey(complement)) {
        return new int[] { indices.get(complement), i };
      }
      indices.putIfAbsent(current, i);
    }

    throw new IllegalArgumentException("Pair not found");
  }

  public static void run() {
    // intentionally empty
  }
}

File: Hashing/python/pairProduct.py
from typing import List, Tuple


def pair_product(numbers: List[int], target_product: int) -> Tuple[int, int]:
    indices: dict[int, int] = {}

    for i, value in enumerate(numbers):
        if value == 0:
            if target_product == 0 and 0 in indices:
                return (indices[0], i)
            if 0 not in indices:
                indices[0] = i
            continue

        if target_product % value != 0:
            if value not in indices:
                indices[value] = i
            continue

        complement = target_product // value
        if complement in indices:
            return (indices[complement], i)
        if value not in indices:
            indices[value] = i

    raise ValueError("Pair not found")
